File: neuron.cpp
//Neuron Class Implemention
// mkrum

#include "Neuron.h"
//default activation functions
double sigmoid(double);
double sigmoidDer(double);

Neuron::Neuron(int size){
    std::random_device r;
    std::default_random_engine gen(r());
    std::uniform_real_distribution<double> dist(0, 1);
    for(int i = 0; i < size; i++){
        weights.push_back(dist(gen));
    }
    error.reserve(size);
    inputs.reserve(size);
    bias = dist(gen);
    _activation = &sigmoid;
    _activationDer = &sigmoidDer;
    stepSize = .1;
}

double Neuron::feedForward(vector<double> ins){
    inputs = ins;
    double sum = 0;
    for(int i = 0; i < weights.size(); i++){
        sum += inputs[i] * weights[i];
    }
    output = _activation(sum + bias);
    return output;
}

vector<double> Neuron::backPropagate(double inError){
    for(int i = 0; i < weights.size(); i++){
        error[i] = inError * _activationDer(output) * weights[i];
        weights[i] -= inError * _activationDer(output) * inputs[i]; 
    }
    bias -= inError * _activationDer(output) * bias;
    return error;
}
 
double sigmoid(double val){
    return 1/(1 + exp(-val));
}

double sigmoidDer(double val){
    return sigmoid(val) * ( 1 - sigmoid(val));
}

   

File: scala/neuron.scala
package neuron

abstract class Neuron(val size : Int, val stepSize : Double){
  val r = new util.Random
  var bias : Double = r.nextFloat
  var weights : Array[Double] = Array.fill(size)(r.nextFloat)
  def activation(x : Double) : Double = x
  def activationDer(y: Double) : Double = y
  var output : Double = _
  var inputs : Array[Double] = _
  def feedForward(ins: Array[Double]) : Double = {
    var sum : Double = 0;
    inputs = ins;
    (inputs, weights).zipped.foreach{ (i,w)  => sum += i*w};
    output = activation(sum + bias);
    output
  }
  def backPropagate(error : Double) : Array[Double] = {
    println(inputs(0));
    (inputs, weights).zipped.map{(i, w) => w  - stepSize * error * activationDer(output) * i};
    bias = bias - stepSize * error * activationDer(output);
    weights.map{w => error * activationDer(output) * w;}
  } 
}
